### PR TITLE
GEOT-4999: Adding support for bbox types to solr datastore.

### PR DESCRIPTION
**Pull request by jdeolive**
Includes introducing the abstraction called "SpatialStrategy" to encapsulate
the handling of different geometry types.

### CODE REVIEW

File: modules/unsupported/solr/src/main/java/org/geotools/data/solr/SolrFeatureReader.java [hunk 1]
```diff
@@ -168,7 +170,9 @@ public SimpleFeature next() throws IOException, IllegalArgumentException,
                 if (type instanceof GeometryDescriptor) {
                     GeometryDescriptor gatt = (GeometryDescriptor) type;
                     if (value != null) {
-                        Geometry geometry = wktReader.read(value.toString());
+                        SolrSpatialStrategy spatialStrategy = SolrSpatialStrategy.createStrategy(gatt);
+                        Geometry geometry = spatialStrategy.decode(value.toString());
```
Comment (aaime) [inline modules/unsupported/solr/src/main/java/org/geotools/data/solr/SolrFeatureReader.java]: Wondering if creating the strategy object at each feature read might be too much overhead.. maybe it should be created once when the reader is instantiated instead?

**Comment by aaime** [comment]
Looking good, minus one potential performance issue in the reader.

**Comment by jdeolive** [comment]
Cool, thanks Andrea. Fixed now and took it one step further making the two strategies singletons since they don't have be instantiated on an attribute by attribute basic, and are thread safe, etc... 

I'll go ahead and merge unless I hear anything back. Thanks for the review.

**Comment by aaime** [comment]
+1

**Pull request merged**